Basic Outline of the Contributors Page #609

sandydays opened the pull request:
The basic outline of the page has been done, including division into sections of what the page could contain and fetching data from the github endpoint of all the contributors and rendering that.

martin-henz commented:
Sandy, can you take a look at this Issue 
https://github.com/source-academy/cadet-frontend/issues/609
and compare. Where are the gaps? Let's reach a consensus first.

geshuming commented on src/components/Contributors.tsx:
```diff
@@ -0,0 +1,167 @@
+}
+
+class Contributors extends Component<{}, IContributorsArrayState> {
+    
```
States suggest the data is mutable during the user's interaction with the page. Can we classify the above structs as `type` instead, for example 
```typescript
type Contributor = {
    ignoreRepos: string[],
    ignoreContributors: string[],
    repos: ReposType[],
    contributors: ContributorsType[]
};

type Repo {
    key: number,
    name: string,
    description: string,
    link: string
};
```

and add the following as constants
```suggestion
private ignoreRepos = ["sicp", "sharedb-ace-backend", "assessments", "tools", "source-academy2"];
private ignoreContributors: ["dependabot[bot]", "dependabot-preview[bot]"],
private repos: Repo[],
private contributors: Contributor[]
```

sandydays replied:
I have changed the structs from interface to type as suggested in my latest commit. Yet to figure out how to change parts of the code if I add ignoreRepos, ignoreContributors, repos, and contributors as constants instead of managing them within the state.

geshuming replied:
You can directly refer to the constants like `this.ignoreRepos` and `this.repos = ...`.

We shouldn't store them as states if they need not be.

sandydays replied:
Yes agreed, I have made the necessary changes in the latest commit.

geshuming commented on src/components/Contributors.tsx:
```diff
@@ -0,0 +1,167 @@
+            const { ignoreRepos } = this.state;
+            const contributorLinks = repos
+                .filter((repo: any) => {
+                    for (const repoName of ignoreRepos) {
```
can be simplified, and try to avoid using the `any` term

```typescript
const contributorLinks = repos
  .filter(repo => !(repo.name in ignoreRepos)
```

openorclose replied:
That's not how the `in` operator works. Use `array.includes(item)` to check if array contains the item.

geshuming replied:
Yep my bad, i was mistaken

sandydays replied:
I have simplified the code within filter in my latest commit.

geshuming commented on src/components/Contributors.tsx:
```diff
@@ -0,0 +1,167 @@
+                        const { ignoreContributors } = this.state;
+                        const contributorList = contributors
+                            .filter((contributor: any) => {
+                                for (const contributorName of ignoreContributors) {
```
same as above `filter` simplification

sandydays commented:
Shuming mentioned:
> Try to avoid the type `any`, the reason we use typescript is so we know what we props and functions an object will have.

I think some places in the code where I cycle through the array and use the type `any` is because at that point I am fetching an object in an array from the api endpoint, where each object has too many key-value pairs + attributes of each object might also change. So, I decided to stick with `any` type in those places. Is that alright?

But you are right, there are some places where I think have used the `any` type but it can be replaced by the types I have already declared at the top of the code. So, will work on fixing those soon.

martin-henz commented:
**team vs contributors**: "team" emphasizes the fact the we are working closely together, and "contributors" emphasizes that people can take part easily. I'd vote for "contributors". That way it's more inviting to newcomers.

martin-henz commented:
**a little stretched...*: I agree. One way to handle is to have two columns, similar to the **Facilitators** of modules in LumiNUS. Also worth considering: we could put the name and number of commits to the right of the picture, instead of under it.

geshuming commented:
> **a little stretched...*: I agree. One way to handle is to have two columns, similar to the Facilitators of modules in LumiNUS. Also worth considering: we could put the name and number of commits to the right of the picture, instead of under it.

[Yes we can use css-grid for this! ](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Auto-placement_in_CSS_Grid_Layout)

sandydays commented:
I have tried to make the popup look better with the latest commit:

1. Made the dialog box wider as well as adjusted the image size of each contributor and the width of their name and number of commits. So, now the contributors are arranged row by row with 4 contributors per row. 
\
@martin-henz suggested that we split it into 2 columns, so that would mean 2 contributors per row? I felt that since there are so many, might as well have 4 per row to make it easier to see more contributors at first glance and have lesser scrolling. I don't know though, which one do you all think is better? 
\
Also, as @martin-henz mentioned, we could put the gitHub username and number of commits to the right of the photo which I don't mind as well. Do we go for this? 

2. Changed "Team" to "Contributors" where necessary as it was voted to be the better option.

With my minimal knowledge of css and sass, I have experimented quite a bit and decided certain things like how many pixels something should be, etc. I am not too sure about issues of how the same thing would appear on different browsers/devices, as well as if the way I have created different classes for different container elements is fine. Any suggestions on that as well later when cleaning code would be helpful as well!

sandydays commented:
@rrtheonlyone I have no idea why the `package-lock.json` file got changed when I made my first commit. I did `git add .` so I didn't realize at that point that the file had changed. I saw it and it seems two lines got deleted for some reason. Added back those two lines, ran the code, everything still worked fine. 

So, what do I do now with regards to that?

geshuming commented:
> @geshuming I used css display: inline-block instead of grid, but I can look into grid as well if you need me to use that!

Yep, we can use css-grid to display 4 contributors per row, and then have the dialog box width fit the content, instead of using a fixed width.

martin-henz commented:
> @sandydays I would recommend against using redux to handle contributor data. You are just complicating the data fetching process without any benefit of using the redux store.
> 
> I would recommend you do an initial fetch when the `Application` performs `componentDidMount()`, store the data as a constant and pass down the prop to the `About` component. No need for redux at all.

Agree. The "single source of truth" is GitHub, here.
